File: get_producturl.py
import os
import sys


class STATE:
    READING_TAG_OPEN = 0
    READING_TAG_JUST_OPEN = 1
    READING_TAG_CLOSE = 2
    READING_STUFF = 3


def got_stuff(tree, stuff):
    if tree and tree[-1] == 'ProductUrl':
        print (stuff)


def get_producturls(f):
    tree, elem, state = [], [], STATE.READING_STUFF

    bu = 'tmp'
    while bu != '':
        bu = f.read(4096)
        for c in bu:
            if state == STATE.READING_STUFF:
                if c == '<':
                    got_stuff(tree, "".join(elem))
                    elem = []
                    state = STATE.READING_TAG_JUST_OPEN
                else:
                    elem.append(c)
            elif state == STATE.READING_TAG_JUST_OPEN:
                if c == '/':
                    state = STATE.READING_TAG_CLOSE
                else:
                    elem.append(c)
                    state = STATE.READING_TAG_OPEN
            elif state == STATE.READING_TAG_OPEN:
                if c == '>':
                    tree.append("".join(elem))
                    elem = []
                    state = STATE.READING_STUFF
                elif c == '/':
                    # skip empty stuff
                    state = STATE.READING_STUFF
                else:
                    elem.append(c)
            elif state == STATE.READING_TAG_CLOSE:
                if c == '>':
                    tree.pop()
                    state = STATE.READING_STUFF


def entry_point(argv):
    if len(argv) != 2:
        print ("Usage: %s <file>" % argv[0])
        return 1

    f = open(argv[1], 'r')
    get_producturls(f)
    f.close()
    return 0


def target(*args):
    return entry_point, None


if __name__ == '__main__':
    entry_point(sys.argv)

File: get_producturl.rs
use std::os;
use std::io::BufferedReader;
use std::io::File;
use std::io;

#[deriving(Show)]
enum State {
    RTagOpen,
    RTagJustOpen,
    RTagClose,
    RMain
}

fn process(name : &str) {

    let path = Path::new(name);
    let mut reader = BufferedReader::new(File::open(&path));
    let mut out = io::stdout();

    let mut state = RMain;
    let mut buf = [0, ..4096];
    let mut tags : Vec<~str> = Vec::new();
    let mut cur_text : Vec<u8> = Vec::new();

    while reader.read(buf).is_ok() {
        for c in buf.iter() {
            match (state, *c as char) {
                (RMain, '<' ) => {
                    if tags.pop() == Some("ProductUrl".to_owned()) {
                        let _ = out.write(cur_text.as_slice());
                        let _ = out.write_line("");
                    }
                    cur_text = Vec::new();
                    state = RTagJustOpen;
                }
                (RMain, _) =>
                    cur_text.push(*c),
                (RTagJustOpen, '/') =>
                    state = RTagClose,
                (RTagJustOpen, _) => {
                    cur_text.push(*c);
                    state = RTagOpen;
                }
                (RTagOpen, '>') => {
                    let v = std::str::from_utf8(cur_text.as_slice()).unwrap().to_owned();
                    tags.push(v);
                    cur_text = Vec::new();
                    state = RMain;
                }
                (RTagOpen, '/') =>
                    state = RMain,
                (RTagOpen, _) =>
                    cur_text.push(*c),
                (RTagClose, '>') => {
                    let _ = tags.pop();
                    state = RMain;
                }
                (RTagClose, _) =>
                    { }
            }
        }
    }
}

fn main() {
    let mut argv = os::args();
    let prog = argv.shift().unwrap();
    let filename = argv.shift();
    match filename {
        None => {
            println!("Usage: {} <filename>", prog);
        }
        Some(name) => {
            process(name);
        }
    }
}
